service/storage_service.cc

```diff
@@ -2347,12 +2347,18 @@ void storage_service::run_replace_ops(std::unordered_set<token>& bootstrap_token
     std::list<gms::inet_address> ignore_nodes = get_ignore_dead_nodes_for_replace(*tmptr);
     // Step 1: Decide who needs to sync data for replace operation
     std::list<gms::inet_address> sync_nodes;
+    locator::host_id local_host_id = _db.local().get_config().host_id;
+    locator::host_id replace_host_id;
     for (const auto& x :_gossiper.get_endpoint_states()) {
         seastar::thread::maybe_yield();
         const auto& node = x.first;
         slogger.debug("replace[{}]: Check node={}, status={}", uuid, node, _gossiper.get_gossip_status(node));
-        if (node != get_broadcast_address() &&
-                node != replace_address &&
+        if (node == replace_address) {
+            auto* host_id_ptr = x.second.get_application_state_ptr(application_state::HOST_ID);
+            if (host_id_ptr) {
+                replace_host_id = locator::host_id(utils::UUID(host_id_ptr->value));
```

Comment:
I think we should fail the operation if we can't retrieve the Host ID of the replaced node.

By this point it should be impossible not to have the Host ID though - if the Host ID was not there, we would've failed earlier, in `prepare_replacement_info`. So it could even be `on_internal_error`.

Reply:
> In general, as a followup, we should move towards specifying the replaced host ID rather than its endpoint (using e.g. replace_host_first_boot config option instead of replace_address_first_boot)

YES PLEASE.

We should deprecate `replace_address_first_boot` ASAP, hopefully in 5.2 already.